### PR TITLE
feature: release on tags

### PR DESCRIPTION
**Pull request by yannicklamprecht**
requires a settings.xml file-content with the configured server as GitHub project secret. Content of the variable `secrets.MAVEN_SETTINGS`

```xml
<settings xmlns="http://maven.apache.org/SETTINGS/1.0.0"
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0
                      https://maven.apache.org/xsd/settings-1.0.0.xsd">
    <servers>
        <server>
            <id>od-releases</id>
            <username>username</username>
            <password>password</password>
        </server>
        <server>
            <id>od-snapshots</id>
            <username>username</username>
            <password>password</password>
        </server>
    </servers>
</settings>
```

### CODE REVIEW

**Comment by lokka30** [comment]
Thank you for the PR! I'd like to hear from @Jikoo @MrIvanPlays @MrNemo64 if possible - I'm too unfamiliar with this to comment on it.

**Comment by MrIvanPlays** [comment]
I used a profile so it doesn't generate sources and javadocs all the time. You may want to move the build plugins back to a profile, probably called "deployment" or sth. 

If repo.mrivanplays.com is going to be used, then ci.mrivanplays.com shall be used, making the releases.yml or whatever it was called irrelevant. I am not willing to put the username and the password of the nexus wherever the fuck I get pleased to, and CodeMC also won't do this, if they respond to the request.

**Comment by MrNemo64** [comment]
> Thank you for the PR! I'd like to hear from @Jikoo @MrIvanPlays @MrNemo64 if possible - I'm too unfamiliar with this to comment on it.

Same here. I didn't even know how to make PRs untill i worked on this repo. All of this is new to me but I'm trying to learn it.

**Comment by portlek** [comment]
there is a simpler, easier and cheaper way but why you just make difficulties for the easiest thing in the world? i don't get it. why we just don't use maven central, why. why we are discussing about the easiest thing for days. i really don't get it. if i know you will accept my release-automation p.r. i will do it.

**Comment by portlek** [comment]
why that obstinacy, i don't get it.

**Comment by yannicklamprecht** [comment]
> I used a profile so it doesn't generate sources and javadocs all the time. You may want to move the build plugins back to a profile, probably called "deployment" or sth.
> 
> If repo.mrivanplays.com is going to be used, then ci.mrivanplays.com shall be used, making the releases.yml or whatever it was called irrelevant. I am not willing to put the username and the password of the nexus wherever the fuck I get pleased to, and CodeMC also won't do this, if they respond to the request.

That's why there are:
1. users
2. roles
3. content selectors
4. user privileges 

I'm not a fan of wasting my own resources when Github Actions is a free to use CI.
In terms of the maven repository, I'm not up to date with the decision of M$ made if Github maven artifact pulling is working without being logged in.

**Comment by MrIvanPlays** [comment]
we didn't do any decision about what to use yet. at first it was about using multiple repositories and multiple cis so if 1 is down, the other one can still function, but then we got to the concern of build numbers and that idea was scrapped. we're waiting on codemc and if we don't get any answer soon, we shall choose either mrivanplays.com or something else.


> That's why there are:
> 
>     1. users
> 
>     2. roles
> 
>     3. content selectors
> 
>     4. user privileges
> 

and how can I create a user in nexus which is able only to deploy to certain repositories (od-releases and od-snapshots) like without logging into the UI located at repo.mrivanplays.com???

**Comment by yannicklamprecht** [comment]
> we didn't do any decision about what to use yet. at first it was about using multiple repositories and multiple cis so if 1 is down, the other one can still function, but then we got to the concern of build numbers and that idea was scrapped. we're waiting on codemc and if we don't get any answer soon, we shall choose either mrivanplays.com or something else.
> 
> > That's why there are:
> > ```
> > 1. users
> > 
> > 2. roles
> > 
> > 3. content selectors
> > 
> > 4. user privileges
> > ```
> 
> and how can I create a user in nexus which is able only to deploy to certain repositories (od-releases and od-snapshots) like without logging into the UI located at repo.mrivanplays.com???

Via a bash script and the Swagger documentation.

1. create a content selector
  post:
  path: `/v1/security/content-selectors`
  body: 
  ```
   {
     "name": "treasury_content_selector_key",
     "description": "Treasury CS",
     "expression": "format == \"maven2\" and path =^ \"/com/github/lokka30/Treasury\""
}
```

2. create a user privilege
    post
    path: `/v1/security/privileges/repository-content-selector`
    body:
    ```
     {
  "name": "treasury_sel_priv",
  "description": "",
  "actions": [
    "ADD",
    "EDIT",
    "READ"
  ],
  "format": null,
  "repository": "maven-snapshots", // *-maven2 for all mave2 repositories
  "contentSelector": "treasury_content_selector_key"
}
```

3. Create role
    post
    path: `/v1/security/roles`
    body:
    ```
    {
  "id": "treasury_role",
  "name": "Treasury Role",
  "description": "Role for pushing Treasury",
  "privileges": [
    "treasury_sel_priv"
  ],
  "roles": []
}
```

4. create user
  post
  path: `/v1/security/users` 
  body:
  ```
{
  "userId": "treasury_user",
  "firstName": "Treasury",
  "lastName": "Push",
  "emailAddress": "<email>",
  "password": "<password>",
  "status": "active",
  "roles": [
    "treasury_role"
  ]
}
```


Hopefully that helps you writing a script for it, requiring only the following as input:

- username
- password
- Email address 
- repositoryFilter
- path to deploy

You welcome.

**Comment by MrIvanPlays** [comment]
@yannicklamprecht creating a user privilege with such a repository is not possible? shall I put the group instead of the repository e.g 

image

**Comment by yannicklamprecht** [comment]
> @yannicklamprecht creating a user privilege with such a repository is not possible? shall I put the group instead of the repository e.g
> 
> image

It should be possible, you probably need to create two privileges and add them both to the role.
Could be somewhat possible using a regex, but not 100% sure.

**Comment by MrIvanPlays** [comment]
Done. What do we do now?

**Comment by lokka30** [comment]
I have heard that Maven Central is a risky place to publish software that uses Bukkit due to the DMCA

**Comment by Jikoo** [comment]
Overall seems good, but I have not taken the time to properly read through the issue.

> I have heard that Maven Central is a risky place to publish software that uses Bukkit due to the DMCA

Central does not like projects that depend on external repos, but that's about it. Bukkit was never DMCAed, Craftbukkit was.

**Comment by lokka30** [comment]
## Lack of understanding
I'd just like to reiterate that like MrNemo I really don't have experience with CI and repository software. The most I've done is have a JitCI page where I manually build each release on GitHub so developers have access to a maven repository. It's not ideal and that's where my knowledge ends. So I'm not able to assist in configuring the CI and repository systems unfortunately - help would be appreciated whenever available.

## Current opinion
* We need a maven repository so that developers can depend on Treasury in their plugins.
* We need a solution that scans each commit to either master or every branch, to check if it builds properly. I assume this is called CI.
* We need a solution that can host development and release builds. Unsure how we will do the versioning, I've created a post in the Discussions section about that.
* Plugin developers must not need to do any extra set-up (e.g. add github token to a file) in order to have their plugin depend on Treasury properly.
* These systems should be automated as much as possible. Of course releases should be manually picked as a version is developed.

Of course my opinion can be changed.

## CodeMC progress
The owner of CodeMC released an announcement a few days ago:

`Gabriele C.` — `06/01/2022`
> Hello everyone, first of all, happy new year! I just wanted to say sorry for my inactivity and the slowness of our support service and resource approvation. The last months were very rough for me and my mental health, but hopefully now things are going better. I promise I'll try to catch up again in the next few days. Thank you all for your patience. 🙂 

They are gradually tending to access requests. I know it's been approx two weeks of nothing happening since we're waiting on solving what systems to use, but I think CodeMC might be the right place for us to put Treasury. They have Jenkins and Nexus, which should have the entire build-and-release system covered to my limited understanding.

***

A poor choice in which systems to use could result in lower adoption, so I'd like that we make the right choice the first time.

Tagging users who may be interested: @Jikoo @MrIvanPlays @yannicklamprecht @portlek @MrNemo64 @NoahvdAa @creatorfromhell

**Comment by MrIvanPlays** [comment]
Yannick's solution is the best in my opinion. Don't waste resources on a CI when github can do that. We can deploy to repo.mrivanplays.com so the api can get adopted and implemented.

**Comment by lokka30** [comment]
> Yannick's solution is the best in my opinion. Don't waste resources on a CI when github can do that. We can deploy to repo.mrivanplays.com so the api can get adopted and implemented.

Awesome. How does the automation work with your repo? And, how should we host our javadocs?

**Comment by MrIvanPlays** [comment]
We can host javadocs via gh pages, I can make it commit stuff after deployment.

**Comment by yannicklamprecht** [comment]
> We can host javadocs via gh pages, I can make it commit stuff after deployment.
Should be basically that: https://github.com/JamesIves/github-pages-deploy-action

**Comment by MrIvanPlays** [comment]
> > We can host javadocs via gh pages, I can make it commit stuff after deployment.
> 
> Should be basically that: https://github.com/JamesIves/github-pages-deploy-action

yep. we just need to exclude the non api modules and we basically have it.

**Comment by lokka30** [comment]
Is there anything else we need to work on in this PR?

I am comfortable waiting another 4 days to see if we are accepted by CodeMC (Gabriel has been looking at requests, we are second from the top at the moment). But if you guys would prefer to use Ivan's repository then we can switch instantly.

**Comment by lokka30** [comment]
What else do we need to work on?

**Comment by MrIvanPlays** [comment]
> Is there anything else we need to work on in this PR?
> What else do we need to work on?

Return the deployment profile back and do some magic to deploy javadocs. 
Also figure out how to do that settings.xml file for the github account.

**Comment by lokka30** [comment]
@yannicklamprecht what else do we need to do to merge this PR?

**Comment by yannicklamprecht** [comment]
> @yannicklamprecht what else do we need to do to merge this PR?

Did @MrIvanPlays created you a deploy account?
You configured the secrets?

After both done, it can be merged.

**Comment by MrIvanPlays** [comment]
> > @yannicklamprecht what else do we need to do to merge this PR?
> 
> 
> 
> Did @MrIvanPlays created you a deploy account?
> 
> You configured the secrets?
> 
> 
> 
> After both done, it can be merged.


Ci.mrivanplays.com is deploying. We don't need the deployment part. 

How is that working? We are bumping to a non snapshot version and it releases or what?

**Comment by yannicklamprecht** [comment]
> > > @yannicklamprecht what else do we need to do to merge this PR?
> > 
> > 
> > Did @MrIvanPlays created you a deploy account?
> > You configured the secrets?
> > After both done, it can be merged.
> 
> Ci.mrivanplays.com is deploying. We don't need the deployment part.
> 
> How is that working? We are bumping to a non snapshot version and it releases or what?

You create a tag and based on that Tag, it creates a release, attaches the artifact to that release and deploys the into the maven repository.

**Comment by MrIvanPlays** [comment]
> > > > @yannicklamprecht what else do we need to do to merge this PR?
> 
> > > 
> 
> > > 
> 
> > > Did @MrIvanPlays created you a deploy account?
> 
> > > You configured the secrets?
> 
> > > After both done, it can be merged.
> 
> > 
> 
> > Ci.mrivanplays.com is deploying. We don't need the deployment part.
> 
> > 
> 
> > How is that working? We are bumping to a non snapshot version and it releases or what?
> 
> 
> 
> You create a tag and based on that Tag, it creates a release, attaches the artifact to that release and deploys the into the maven repository.
> 
> 

We don't need it deployed anymore. Could you please update it so it matches with this? And also the whole lots of pom changes we had? Thanks.

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Set up JDK
+        uses: actions/setup-java@v2
+        with:
+          distribution: adopt
+          java-version: 17
```
Comment (lokka30) [inline .github/workflows/release.yml:19]: Is 17 intended? We're using 8. I know, it sucks.

File: pom.xml [hunk 2]
```diff
@@ -144,110 +240,6 @@
                     <url>https://repo.mrivanplays.com/repository/od-snapshots/</url>
                 </snapshotRepository>
             </distributionManagement>
-            <build>
-                <defaultGoal>clean package</defaultGoal>
-                <finalName>
-                    ${project.artifactId}-${plugin.version}-${git.commit.id.abbrev}-${plugin.specificationVersion}
```
Comment (lokka30) [inline pom.xml:150]: Is this `finalName` specified anywhere else?

Comment (MrIvanPlays) [reply pom.xml:150]: That has nothing to do with this scope of this pr

**Comment by MrIvanPlays** [comment]
Why are we moving the sources and javadocs plugins from the profile?

File: pom.xml (hunk 2 above)
Comment (lokka30) [reply pom.xml:150]: I noticed it was removed, is the functionality still available?

Comment (MrIvanPlays) [reply pom.xml:150]: Are you blind or sth??

Comment (lokka30) [reply pom.xml:150]: Couldn't see an equivalent when I was browsing the files of this PR.

Comment (MrIvanPlays) [reply pom.xml:150]: You are blind confirmed.

**Comment by yannicklamprecht** [comment]
> Why are we moving the sources and javadocs plugins from the profile? 

Easy answer to that. We need all of them outside deployment. The only thing that should be different is the distribution repository.

The other way you would have a big duplicated profile section when deploying to the other repository.

The more important reason is that we need that on github release.

File: .github/workflows/release.yml (hunk 1 above)
Comment (yannicklamprecht) [reply .github/workflows/release.yml:19]: Yes intended. Don't like using an old jdk, when the maven compile target level is defined anyway.

**Comment by MrIvanPlays** [comment]
doesn't it currently release every time it builds? I have these concerns due to the part 

```yaml
on:
  push:
    tags:
      - '*'
```

**Comment by yannicklamprecht** [comment]
> > > doesn't it currently release every time it builds? I have these concerns due to the part
> > > ```yaml
> > > on:
> > >   push:
> > >     tags:
> > >       - '*'
> > > ```
> > 
> > Pretty certain this only runs if the push is for a tag specifically, however, if we want to roll this into a more generic CI run I have a functional example here: https://github.com/Jikoo/OpenInv/blob/abd529be6636b8a21787e3ed799a30a42fb6f0b9/.github/workflows/ci.yml The release job [only runs on pushes for tags](https://github.com/Jikoo/OpenInv/blob/abd529be6636b8a21787e3ed799a30a42fb6f0b9/.github/workflows/ci.yml#L53) while the general build job always runs.
> 
> I don't understand github configs at all. Does this solution suit us or no @Jikoo ? If it doesn't I'd like you to modify it so it does. Or yannick can do that. Thanks. 

It's the same on gitlab-ci. See no reason except 20 years old jenkins why it should be different.

When the pushed commit is tagged, run this workflow. The star is for all tags, you can pass a different schema for different actions.
Eg. If you want a pre release you can define a different regex for that.
\d+\\.\d+\\.\d-pre

So it runs in when tags like that are pushed
1.1.1-pre
1.1.2-pre
But not on
1.1.1
1.1.2-release

**Comment by Jikoo** [comment]
> I don't understand github configs at all. Does this solution suit us or no @Jikoo ? If it doesn't I'd like you to modify it so it does. Or yannick can do that. Thanks.

Sorry, I'm a bit out of the loop on our actual direction here, trying to catch up on over a week of reading - I haven't been doing anything more than closing issues that require more than a single reference while on vacation.

I guess it depends on what we want from the action. Do we want GH to control building download releases for user consumption with the CI being separate and building releases for Maven? It shouldn't matter, but files via Maven would be sourced differently, technically.
If the answer is yes, Lokka needs to set up a repo secret as described by the initial PR description using a username and password provided by you for the account(s) deploying the build. Only releases would currently be pushed, but it could be adapted to also push snapshots to the snapshot repo.
If the answer is no (and you don't mind devoting your CI's resources to rebuilding something that was already built with GH's processing power), the settings changes should be scrapped and the build will "deploy" to GH as usual by creating a release after the build completes.

My method uses an existing CI job with a few modifications - the config I linked runs exclusively on actual pushes of tags, while the main build job also runs on PRs. If a tag is created, it first waits for the regular build to complete, then downloads the artifacts created and adds them to a draft release for manual polish. It's more in line with the no answer above, but instead of creating a separate job it uses the files created by the CI run. I specifically do it that way because I haven't bothered setting up an external CI, but it could be compacted further (and save GH some storage space) by not attaching files to the run at all.

**Comment by MrIvanPlays** [comment]
> > I don't understand github configs at all. Does this solution suit us or no @Jikoo ? If it doesn't I'd like you to modify it so it does. Or yannick can do that. Thanks.
> 
> Sorry, I'm a bit out of the loop on our actual direction here, trying to catch up on over a week of reading - I haven't been doing anything more than closing issues that require more than a single reference while on vacation.
> 
> I guess it depends on what we want from the action. Do we want GH to control building download releases for user consumption with the CI being separate and building releases for Maven? It shouldn't matter, but files via Maven would be sourced differently, technically. If the answer is yes, Lokka needs to set up a repo secret as described by the initial PR description using a username and password provided by you for the account(s) deploying the build. Only releases would currently be pushed, but it could be adapted to also push snapshots to the snapshot repo. If the answer is no (and you don't mind devoting your CI's resources to rebuilding something that was already built with GH's processing power), the settings changes should be scrapped and the build will "deploy" to GH as usual by creating a release after the build completes.
> 
> My method uses an existing CI job with a few modifications - the config I linked runs exclusively on actual pushes of tags, while the main build job also runs on PRs. If a tag is created, it first waits for the regular build to complete, then downloads the artifacts created and adds them to a draft release for manual polish. It's more in line with the no answer above, but instead of creating a separate job it uses the files created by the CI run. I specifically do it that way because I haven't bothered setting up an external CI, but it could be compacted further (and save GH some storage space) by not attaching files to the run at all.

I don't mind computing 24 seconds of a machine that runs for 0.010 € per hour. 

What we want is GitHub to act like a jenkins instance of some sort, but only build rather than build and do other stuff (just like ci.mrivanplays.com does - it builds treasury and deploys to nexus and deploys javadocs) and the jars that built to be downloadable. Or at least that's what I understood from Lokka's words. 

How do tags work? Say I have a commit which shall be the last commit for version x and I need that version released automatically. What shall I do? In jenkins it's easy and it's already been setup (kind of).

**Comment by Jikoo** [comment]
> > I guess it depends on what we want from the action. Do we want GH to control building download releases for user consumption with the CI being separate and building releases for Maven?
>
> What we want is GitHub to act like a jenkins instance of some sort, but only build rather than build and do other stuff (just like ci.mrivanplays.com does - it builds treasury and deploys to nexus and deploys javadocs) and the jars that built to be downloadable. Or at least that's what I understood from Lokka's words.

So that's a no to that question then. In that case, the PR needs to be modified to not deploy using Maven and only "deploy" by creating a GitHub release with the appropriate jars attached.

> How do tags work? Say I have a commit which shall be the last commit for version x and I need that version released automatically. What shall I do? In jenkins it's easy and it's already been setup (kind of).

Tags are a separate thing. For example, `git tag 1.2.3`, then `git push <remote> 1.2.3` would create and push the tag 1.2.3. If an action runs only on tag pushes, it will only run when a new tag is pushed and it will run with head set to the commit attached to the tag. It's still a push, just not a push of new commits. Tags can be created at any point separately by specifying the commit hash. GH actions do not run on existing tags despite new pushes attempting to add them, which is good for when people push all tags despite it generally being a bad idea.

**Comment by lokka30** [comment]
I don't mind how any of this is achieved, so long:

* automation is achieved wherever feasible
* developers can depend on Treasury via a Maven repository that requires no additional set-up (like GH Packages does)
* Everyone can download a compiled build of each commit (except for the `[ci ignore], or something similar, exemption used on Ivan's CI).
* A CI can tell us if there were any issues in compiling a commit.

The only things I understand are out of Treasury's scope are:

* GH Packages - requires additional setup for other developers, don't want that at all
* Maven Central - bigger hassle to use, but I can consider it if it is requested

What would be ideal, is to use as much of GitHub's CI power as we can, and as little of Ivan's, CodeMC's, and so on.

**Comment by MrIvanPlays** [comment]
> I don't mind how any of this is achieved, so long:
> 
>     * automation is achieved wherever feasible
> 
>     * developers can depend on Treasury via a Maven repository that requires no additional set-up (like GH Packages does)
> 
>     * everyone can download a compiled build of each commit (except for the `[ci skip]`-prefixed ones, an exemption used on Ivan's CI).
> 
>     * a CI system can tell us if there were any issues in compiling a commit.
> 
> 
> The only things that I know for certain aren't good choices for the project are:
> 
>     * GH Packages - requires additional setup for other developers, don't want that at all.
> 
>     * Maven Central - bigger hassle to use, but I can consider it if it is requested.
> 
> 
> What would be ideal, is to use as much of GitHub's CI power as we can, and as little of Ivan's, CodeMC's, and so on.

In other words, this PR does or does not suit us?

**Comment by lokka30** [comment]
> > I don't mind how any of this is achieved, so long:
> > ```
> > * automation is achieved wherever feasible
> > 
> > * developers can depend on Treasury via a Maven repository that requires no additional set-up (like GH Packages does)
> > 
> > * everyone can download a compiled build of each commit (except for the `[ci skip]`-prefixed ones, an exemption used on Ivan's CI).
> > 
> > * a CI system can tell us if there were any issues in compiling a commit.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The only things that I know for certain aren't good choices for the project are:
> > ```
> > * GH Packages - requires additional setup for other developers, don't want that at all.
> > 
> > * Maven Central - bigger hassle to use, but I can consider it if it is requested.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What would be ideal, is to use as much of GitHub's CI power as we can, and as little of Ivan's, CodeMC's, and so on.
> 
> In other words, this PR does or does not suit us?

I'm too inexperienced with the systems involved in this PR to make a comment. From what I do know, I've tried to clarify what i think is best for this project.

**Comment by MrIvanPlays** [comment]
> I'm too inexperienced with the systems involved in this PR to make a comment. From what I do know, I've tried to clarify what i think is best for this project.

What I've understood is that this will create a release automatically when you do these 2 commands

```
git tag <tag, for an example 1.2.3>
git push <branch> <tag>
```
Do you agree with it or not?

**Comment by lokka30** [comment]
> > I'm too inexperienced with the systems involved in this PR to make a comment. From what I do know, I've tried to clarify what i think is best for this project.
> 
> What I've understood is that this will create a release automatically when you do these 2 commands
> 
> ```
> git tag <tag, for an example 1.2.3>
> git push <branch> <tag>
> ```
> 
> Do you agree with it or not?

That sounds awesome to me. Does it do the automatic changelog generation? All good if not, bells and whistles on that level are definitely not needed but would be a cool feature, besides.

**Comment by MrIvanPlays** [comment]
> Does it do the automatic changelog generation?

I don't know but I'll merge it in a min as I see you agree on having it.

**Pull request merged**